### ISSUE TITLE
OSU osu_bibw fails with assertion due to auto data progress

### ISSUE THREAD

**Issue by sungeunchoi** [ofi-cray/libfabric-cray#676]
```
osu_bibw: pml_cm_recvreq.c:53: mca_pml_cm_recv_request_completion: Assertion `0 == ((mca_pml_cm_thin_recv_request_t*)base_request)->req_base.req_pml_complete' failed.
```

This could be a similar problem to #668

**Comment by hppritcha** [comment]
It appears that if this is not a bug in ompi, it might be some odd end-case with our tag matching engine.  The error indicates that a recv request appears to complete twice.

Its not clear this happens only with progress threads present.

**Comment by hppritcha** [comment]
Testing with MPICH rather than Open MPI, every 30-40 runs or so, the osu_bibw hangs after 4096 message size.  So this is most likely a problem with the provider, and not something particular to Open MPI.

**Comment by hppritcha** [comment]
there seems to be some issue with accessing of singly linked lists in a multi-threaded environment.  After lots of runs, I got a meltdown (seg fault) in `slist_remove_head`.  A segv was hit because the head was null, but the `list->tail` element still pointed to something, so the code ended up attempting to dereference `item->next` when `item` is NULL.  I modified the `slist_insert_tail` to NULL ify `item->next` to NULL.  This works fine if no progress thread, but just adding this single line causes frequent hangs when data progress thread is enabled.

**Comment by hppritcha** [comment]
Okay, found a problem with multi-threaded.  In `__nic_tx_progress` we are releasing the nic lock too early in the case of completer functions that invoke `GNI_SmsgRelease`.   With the current code, sometimes a SMSG message gets dequeued twice.

**Comment by hppritcha** [comment]
this has been fixed with commit #679
